packages/js/src/ui/helpers/utils.ts

```diff
@@ -51,41 +51,173 @@ export function createClassAndRuleFromCssString(classNameSet: Set<string>, style
-  const rules = [];
-  for (let i = 0; i < shades.length; i += 1) {
-    const shade = shades[i];
-    const cssVariableSolidRule = `.${props.id} { --nv-${props.key}-${shade}: oklch(from ${props.color} calc(l - ${
-      (shade - 500) / 1000
-    }) c h); }`;
-    rules.push(cssVariableSolidRule);
-  }
+export function generateSolidShadeRulesFromColor({ color, key, id }: { color: string; key: string; id: string }) {
+  const rules: string[] = [];
+
+  // Adjusted scaling function that works better with white/light colors
+  const adjustLightness = (factor: number) => {
+    if (factor >= 0) {
+      // For lighter shades, approach 1 (white) more directly
+      return `min(1, calc(l + ${factor} * (1 - l)))`;
+    } else {
+      // For darker shades, scale down from current lightness
+      return `max(0, calc(l * (1 + ${factor})))`;
+    }
+  };
+
+  // Adjusted factors for better distribution
+  const lightnessOffsets: Record<number, string> = {
+    25: adjustLightness(0.475), // Very light
+    50: adjustLightness(0.45), // Light
+    100: adjustLightness(0.4), // Slightly light
+    200: adjustLightness(0.3),
+    300: adjustLightness(0.2),
+    400: adjustLightness(0.1),
+    500: 'l', // Base color (unchanged)
+    600: adjustLightness(-0.1),
+    700: adjustLightness(-0.2),
+    800: adjustLightness(-0.3),
+    900: adjustLightness(-0.4), // Very dark
+  };
+
+  shades.forEach((shade) => {
+    const newLightness = lightnessOffsets[shade];
+    const cssVariableRule = `.${id} { --nv-${key}-${shade}: oklch(from ${color} ${newLightness} c h); }`;
+    rules.push(cssVariableRule);
+  });
 
   return rules;
 }
 
-export function generatesAlphaShadesFromColor(props: { color: string; key: string; id: string }) {
-  const rules = [];
-  for (let i = 0; i < shades.length; i += 1) {
-    const shade = shades[i];
-    const cssVariableAlphaRule = `.${props.id} { --nv-${props.key}-${shade}: oklch(from ${props.color} l c h / ${
-      shade / 1000
-    }); }`;
+export function generateAlphaShadeRulesFromColor({ color, key, id }: { color: string; key: string; id: string }) {
+  const rules: string[] = [];
+  const alphaMap = {
+    25: 0.025,
+    50: 0.05,
+    100: 0.1,
+    200: 0.2,
+    300: 0.3,
+    400: 0.4,
+    500: 0.5,
+    600: 0.6,
+    700: 0.7,
+    800: 0.8,
+    900: 0.9,
+  };
+
+  Object.entries(alphaMap).forEach(([shade, alpha]) => {
+    const cssVariableAlphaRule = `.${id} { --nv-${key}-${shade}: oklch(from ${color} l c h / ${alpha}); }`;
     rules.push(cssVariableAlphaRule);
+  });
+
+  return rules;
+}
+
+function getBrowserFontSize(): number {
+  if (typeof window === 'undefined') return 16;
+
+  const rootFontSize = window.getComputedStyle(document.documentElement).fontSize;
+
+  return parseFloat(rootFontSize);
+}
+
+function parseSizeToRem(size: string): number {
+  const browserFontSize = getBrowserFontSize();
+  const match = size.match(/^([\d.]+)(px|rem|em)$/);
```

Comment:
Should we also parse `pt` and `ex`? https://developer.mozilla.org/en-US/docs/Web/CSS/font-size

Reply:
I've updated the logic. We no longer calculate anything. We use calc under the hood.